hibernate-core/src/main/java/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java

```diff
@@ -1407,6 +1416,17 @@ public void pruneForSubclasses(TableGroup tableGroup, Map<String, EntityNameUse>
 		else {
 			tableReferenceJoins.removeIf( join -> !retainedTableReferences.contains( join.getJoinedTableReference() ) );
 		}
+
+		if ( needsTreatDiscriminator ) {
+			// We have to add the discriminator condition to the root table join to make sure
+			// we're only retrieving the correct subtype records from the shared joined table
+			final String discriminatorFragment = getPrunedTableExpression(
+					entityNameUses,
+					metamodel,
+					tableGroup.getPrimaryTableReference().getIdentificationVariable()
+			);
+			tableReferenceJoins.get( 0 ).applyPredicate( new SqlFragmentPredicate( discriminatorFragment ) );
```

Comment:
Do you know for sure that the first join is always the root table join?

Reply:
From what I could understand, the primary table reference is the root table and the joins "add" the joined-subclasses tables - so the first join in the list should always start from the root table.

Reply:
There might be a case where there is no table reference join i.e. when pruning all subclass tables. Here a sample query: `select 1 from RootEntity r where r.rootEntityAttribute = 1`

Reply:
Right, I guess this is a problem only if `RootEntity` is explicitly treated When table reference joins are empty we could resort to the subquery-style filter: `select 1 from (select * from RootEntity t where t.disc_col = '1') r where r.rootEntityAttribute = 1`

Reply:
I've updated the logic to always substitute the root table expression with a subquery containing the discriminator predicate.

I've also confirmed that the tableGroup's primary table reference might not be the root table, and that the first join is not always the one containing it. Also, I've fixed an issue where for explicitly treated joins we were removiong the root table's join causing an error.